Sources/FoundationEssentials/String/String+Path.swift

```diff
@@ -137,6 +137,72 @@ extension String {
     
     #if !NO_FILESYSTEM
     internal static func homeDirectoryPath(forUser user: String? = nil) -> String {
+#if os(Windows)
+        if let user {
+            return user.withCString(encodedAs: UTF16.self) { pwszUserName in
+                var cbSID: DWORD = 0
+                var cchReferencedDomainName: DWORD = 0
+                var eUse: SID_NAME_USE = SidTypeUnknown
+                guard LookupAccountNameW(nil, pwszUserName, nil, &cbSID, nil, &cchReferencedDomainName, &eUse) else {
+                    return ""
+                }
+
+                return withUnsafeTemporaryAllocation(of: CChar.self, capacity: Int(cbSID)) { pSID in
+                    return withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: Int(cchReferencedDomainName)) { pwszReferencedDomainName in
+                        guard LookupAccountNameW(nil, pwszUserName, pSID.baseAddress, &cbSID, pwszReferencedDomainName.baseAddress, &cchReferencedDomainName, &eUse) else {
+                            return ""
+                        }
+
+                        var pwszSID: LPWSTR? = nil
+                        guard ConvertSidToStringSidW(pSID.baseAddress, &pwszSID) else { return "" }
+
+                        return #"SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList\#\(String(decodingCString: pwszSID!, as: UTF16.self))"#.withCString(encodedAs: UTF16.self) { pwszKeyPath in
+                            return "ProfileImagePath".withCString(encodedAs: UTF16.self) { pwszKey in
+                                var cbData: DWORD = 0
+                                guard RegGetValueW(HKEY_LOCAL_MACHINE, pwszKeyPath, pwszKey, RRF_RT_REG_SZ, nil, nil, &cbData) == ERROR_SUCCESS else {
+                                    return ""
+                                }
+                                return withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: Int(cbData)) { pwszData in
+                                    guard RegGetValueW(HKEY_LOCAL_MACHINE, pwszKeyPath, pwszKey, RRF_RT_REG_SZ, nil, pwszData.baseAddress, &cbData) == ERROR_SUCCESS else {
+                                        return ""
+                                    }
+                                    return String(decodingCString: pwszData.baseAddress!, as: UTF16.self)
+                                }
+                            }
+                        }
+
+                    }
+                }
+            }
+        }
+
+        var hToken: HANDLE? = nil
+        guard OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY, &hToken) else {
+            return "USERPROFILE".withCString(encodedAs: UTF16.self) { pwszVariable in
```

Comment:
I think we probably want to take the `USERPROFILE` path as a potential fallback in all of the failure cases and not just this one, right? Should we factor this out into a function that we can call from each failure case?

Reply:
Well, the failure case itself can fail too. So we can only go so far with that.

Reply:
Agreed, if looking up this environment value fails then we need to have something to fallback to (like the empty string), but it feels somewhat arbitrary to only fallback to this `USERPROFILE` value in this one case. In the non-Windows path, we always check `HOME` before falling back to `/var/empty` if `HOME` isn't set, so it seems like if we want consistent behavior we should always check `USERPROFILE` before falling back to something like the empty string, thoughts?

Reply:
Yeah, always falling back to `USERPROFILE` if no user is specified seems reasonable.